### PR TITLE
[X86] Support APX promoted RAO-INT and MOVBE instructions

### PR DESCRIPTION
**Pull request by XinWang10**
R16-R31 was added into GPRs in https://github.com/llvm/llvm-project/pull/70958,
This patch supports the  promoted RAO-INT and MOVBE instructions in EVEX space.

RFC: https://discourse.llvm.org/t/rfc-design-for-apx-feature-egpr-and-ndd-support/73031/4

### CODE REVIEW

**Review by KanRobert — COMMENTED** [review]
Need to add pattern for MOVBErr_EVEX, which has the same sematic as "NDD BSWAP"

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 1]
```diff
@@ -229,7 +229,7 @@ def PUSHA16  : I<0x60, RawFrm, (outs), (ins), "pushaw", []>,
                OpSize16, Requires<[Not64BitMode]>;
 }
 
-let Constraints = "$src = $dst", SchedRW = [WriteBSWAP32] in {
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:232]: We should use NoNDD here from semantic.

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 2]
```diff
@@ -1121,6 +1121,62 @@ let Predicates = [HasMOVBE] in {
...
+                        "movbe{l}\t{$src, $dst|$dst, $src}",
+                        [(set GR32:$dst, (bswap GR32:$src))]>,
+                      EVEX, NoCD8, T_MAP4;
+def MOVBE64rr_EVEX : RI<0x61, MRMDestReg, (outs GR64:$dst), (ins GR64:$src),
+                        "movbe{q}\t{$src, $dst|$dst, $src}",
+                        [(set GR64:$dst, (bswap GR64:$src))]>,
+                      EVEX, NoCD8, T_MAP4;
+
+def MOVBE16rr_EVEX_REV : I<0x60, MRMSrcReg, (outs GR16:$dst), (ins GR16:$src),
+                            "movbe{w}\t{$src, $dst|$dst, $src}", []>,
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1170]: Reverse encoding should inherits class DisassemblyOnly.

And put the decoding tests at ./llvm/test/MC/Disassembler/X86/apx/reverse-encoding.txt

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 3]
```diff
@@ -1090,7 +1090,7 @@ def ARPL16mr : I<0x63, MRMDestMem, (outs), (ins i16mem:$dst, GR16:$src),
 //===----------------------------------------------------------------------===//
 // MOVBE Instructions
 //
-let Predicates = [HasMOVBE] in {
+let Predicates = [HasMOVBE, NoEGPR] in {
   let SchedRW = [WriteALULd] in {
   def MOVBE16rm : I<0xF0, MRMSrcMem, (outs GR16:$dst), (ins i16mem:$src),
                     "movbe{w}\t{$src, $dst|$dst, $src}",
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1096]: Rewrite MOVBE with `ITy`

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 4]
```diff
@@ -25,21 +25,30 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
 def X86rao_and  : SDNode<"X86ISD::AAND", SDTRAOBinaryArith,
                          [SDNPHasChain, SDNPMayStore, SDNPMemOperand]>;
 
-multiclass RAOINT_BASE<string OpcodeStr> {
+multiclass RAOINT_BASE<string OpcodeStr, string Suffix = ""> {
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:28]: OpcodeStr -> m
Suffix -> suffix

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 5]
```diff
@@ -25,21 +25,30 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
...
 
   let Predicates = [HasRAOINT, In64BitMode] in
-    def 64mr : I<0xfc, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
-                 !strconcat("a", OpcodeStr, "{q}\t{$src, $dst|$dst, $src}"),
-                 [(!cast<SDNode>("X86rao_" # OpcodeStr) addr:$dst, GR64:$src)]>,
-               Sched<[WriteALURMW]>, REX_W;
+    def 64mr#Suffix : I<0xfc, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
+                        !strconcat("a", OpcodeStr, "{q}\t{$src, $dst|$dst, $src}"),
+                        [(!cast<SDNode>("X86rao_" # OpcodeStr) addr:$dst, GR64:$src)]>,
+                      Sched<[WriteALURMW]>, REX_W;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:39]: src -> src1

{$src, $dst|$dst, $src} -> binop_args

File: llvm/test/TableGen/x86-fold-tables.inc [hunk 6]
```diff
@@ -349,6 +349,9 @@ static const X86FoldTableEntry Table0[] = {
   {X86::MOV8rr_NOREX, X86::MOV8mr_NOREX, TB_FOLDED_STORE|TB_NO_REVERSE},
   {X86::MOVAPDrr, X86::MOVAPDmr, TB_FOLDED_STORE|TB_NO_REVERSE|TB_ALIGN_16},
   {X86::MOVAPSrr, X86::MOVAPSmr, TB_FOLDED_STORE|TB_NO_REVERSE|TB_ALIGN_16},
+  {X86::MOVBE16rr_EVEX, X86::MOVBE16mr_EVEX, TB_FOLDED_STORE},
```
Comment (KanRobert) [inline llvm/test/TableGen/x86-fold-tables.inc:352]: Need to figure out why MOVBE16rr is not here while MOVBE16rr_EVEX is

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 7]
```diff
@@ -1090,35 +1090,43 @@ def ARPL16mr : I<0x63, MRMDestMem, (outs), (ins i16mem:$dst, GR16:$src),
...
+
+multiclass Movberr<X86TypeInfo t> {
+  def rr_EVEX : ITy<0x61, MRMDestReg, t, (outs t.RegClass:$dst),
+                    (ins t.RegClass:$src1), "movbe", unaryop_ndd_args,
+                    [(set t.RegClass:$dst, (bswap t.RegClass:$src1))]>,
+                EVEX, T_MAP4;
+  def rr_EVEX_REV : ITy<0x60, MRMSrcReg, t, (outs t.RegClass:$dst),
+                        (ins t.RegClass:$src1), "movbe", unaryop_ndd_args, []>,
+                    EVEX, T_MAP4, DisassembleOnly;
+}
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1125]: Remove _EVEX for rr variant b/c there is no non-EVEX one.

Need to compress MOVEBE_rr to BSWAP if src1 and dst are same, and add test in compress-evex.mir

File: llvm/test/TableGen/x86-fold-tables.inc (hunk 6 above)
Comment (XinWang10) [reply llvm/test/TableGen/x86-fold-tables.inc:352]: We don't have MOVBE16rr for legacy encoding.

Comment (KanRobert) [reply llvm/test/TableGen/x86-fold-tables.inc:352]: Ah, I see

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 8]
```diff
@@ -1090,35 +1090,43 @@ def ARPL16mr : I<0x63, MRMDestMem, (outs), (ins i16mem:$dst, GR16:$src),
...
+
+multiclass Movberr<X86TypeInfo t> {
+  def rr : ITy<0x61, MRMDestReg, t, (outs t.RegClass:$dst),
+               (ins t.RegClass:$src1), "movbe", unaryop_ndd_args,
+               [(set t.RegClass:$dst, (bswap t.RegClass:$src1))]>,
+           EVEX, T_MAP4;
+  def rr_REV : ITy<0x60, MRMSrcReg, t, (outs t.RegClass:$dst),
+                   (ins t.RegClass:$src1), "movbe", unaryop_ndd_args, []>,
+               EVEX, T_MAP4, DisassembleOnly;
+}
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1125]: Why not `UnaryOpR_R`?

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 9]
```diff
@@ -1090,35 +1090,43 @@ def ARPL16mr : I<0x63, MRMDestMem, (outs), (ins i16mem:$dst, GR16:$src),
...
+  let SchedRW = [WriteALULd] in
+  def rm#suffix : ITy<o, MRMSrcMem, t, (outs t.RegClass:$dst),
+                      (ins t.MemOperand:$src1), "movbe", unaryop_ndd_args,
+                      [(set t.RegClass:$dst, (bswap (t.LoadNode addr:$src1)))]>;
+  let SchedRW = [WriteStore] in
+  def mr#suffix : ITy<!add(o, 1), MRMDestMem, t, (outs),
+                      (ins t.MemOperand:$dst, t.RegClass:$src1),
+                      "movbe", unaryop_ndd_args,
+                      [(store (bswap t.RegClass:$src1), addr:$dst)]>;
+}
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1103]: Why not `UnaryOpR_M`

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 10]
```diff
@@ -25,21 +25,30 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
 def X86rao_and  : SDNode<"X86ISD::AAND", SDTRAOBinaryArith,
                          [SDNPHasChain, SDNPMayStore, SDNPMemOperand]>;
 
-multiclass RAOINT_BASE<string OpcodeStr> {
+multiclass RAOINT_BASE<string m, string suffix = ""> {
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:28]: RAOINT_BASE -> RaoInt

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 11]
```diff
@@ -25,21 +25,30 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
...
   let Predicates = [HasRAOINT] in
-    def 32mr : I<0xfc, MRMDestMem, (outs), (ins i32mem:$dst, GR32:$src),
-                 !strconcat("a", OpcodeStr, "{l}\t{$src, $dst|$dst, $src}"),
-                 [(!cast<SDNode>("X86rao_" # OpcodeStr) addr:$dst, GR32:$src)]>,
-               Sched<[WriteALURMW]>;
+    def 32mr#suffix : I<0xfc, MRMDestMem, (outs), (ins i32mem:$src1, GR32:$src2),
+                        !strconcat("a", m, "{l}\t", binop_args),
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:31]: Use `BinOpMR_M`. The schedule should be corrected to
`WriteALURMW, ReadDefault, ReadDefault, ReadDefault, ReadDefault, ReadDefault`. You can create a separate PR for this.

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrRAOINT.td:31]: What's the difference between [WriteALURMW] and [WriteALURMW, ReadDefault, ReadDefault, ReadDefault, ReadDefault, ReadDefault]

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrRAOINT.td:31]: You will find the difference if you check the instruction in MCA test

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 12]
```diff
@@ -1090,35 +1090,43 @@ def ARPL16mr : I<0x63, MRMDestMem, (outs), (ins i16mem:$dst, GR16:$src),
...
-  def MOVBE64mr : RI<0xF1, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
-                     "movbe{q}\t{$src, $dst|$dst, $src}",
-                     [(store (bswap GR64:$src), addr:$dst)]>,
-                     T8;
-  }
+multiclass Movbe<bits<8> o, X86TypeInfo t, string suffix = ""> {
+  let SchedRW = [WriteALULd] in
+  def rm#suffix : ITy<o, MRMSrcMem, t, (outs t.RegClass:$dst),
+                      (ins t.MemOperand:$src1), "movbe", unaryop_ndd_args,
+                      [(set t.RegClass:$dst, (bswap (t.LoadNode addr:$src1)))]>;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1097]: Usually we should indent 2 more spaces for the `def`.

To reduce the indentation, put `Sched<WriteALULd>` after the ITy.

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 13]
```diff
@@ -1090,35 +1090,43 @@ def ARPL16mr : I<0x63, MRMDestMem, (outs), (ins i16mem:$dst, GR16:$src),
...
+  let SchedRW = [WriteStore] in
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:1098]: ditto

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 14]
```diff
@@ -25,21 +25,26 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
...
 
-defm AADD : RAOINT_BASE<"add">, T8;
-defm AAND : RAOINT_BASE<"and">, T8, PD;
-defm AOR  : RAOINT_BASE<"or" >, T8, XD;
-defm AXOR : RAOINT_BASE<"xor">, T8, XS;
+let Predicates = [HasRAOINT, NoEGPR] in {
+defm AADD : RaoInt<"add">, T8;
+defm AAND : RaoInt<"and">, T8, PD;
+defm AOR  : RaoInt<"or" >, T8, XD;
+defm AXOR : RaoInt<"xor">, T8, XS;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:42]: Indent

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 15]
```diff
@@ -25,21 +25,26 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
...
+    def 32mr#suffix : BinOpMR<0xfc, "a"#m, binop_args, Xi32, (outs),
+                        [(!cast<SDNode>("X86rao_" # m) addr:$src1, GR32:$src2)]>;
 
-  let Predicates = [HasRAOINT, In64BitMode] in
-    def 64mr : I<0xfc, MRMDestMem, (outs), (ins i64mem:$dst, GR64:$src),
-                 !strconcat("a", OpcodeStr, "{q}\t{$src, $dst|$dst, $src}"),
-                 [(!cast<SDNode>("X86rao_" # OpcodeStr) addr:$dst, GR64:$src)]>,
-               Sched<[WriteALURMW]>, REX_W;
+    def 64mr#suffix : BinOpMR<0xfc, "a"#m, binop_args, Xi64, (outs),
+                        [(!cast<SDNode>("X86rao_" # m) addr:$src1, GR64:$src2)]>, REX_W;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:34]: Why not `BinOpMR_M`? Then you can remove `let SchedRW = [WriteALURMW, ReadDefault, ReadDefault, ReadDefault, ReadDefault, ReadDefault]` b/c it's default value

File: llvm/lib/Target/X86/X86InstrRAOINT.td [hunk 16]
```diff
@@ -25,21 +25,26 @@ def X86rao_xor  : SDNode<"X86ISD::AXOR", SDTRAOBinaryArith,
...
+defm AAND : RaoInt<"and">, T8, PD;
+defm AOR  : RaoInt<"or" >, T8, XD;
+defm AXOR : RaoInt<"xor">, T8, XS;
+}
+
+let Predicates = [HasRAOINT, HasEGPR, In64BitMode] in {
+defm AADD : RaoInt<"add", "_EVEX">, EVEX, T_MAP4;
+defm AAND : RaoInt<"and", "_EVEX">, EVEX, T_MAP4, PD;
+defm AOR  : RaoInt<"or",  "_EVEX">, EVEX, T_MAP4, XD;
+defm AXOR : RaoInt<"xor", "_EVEX">, EVEX, T_MAP4, XS;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrRAOINT.td:49]: Indent

File: llvm/lib/Target/X86/X86InstrMisc.td (hunk 7 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrMisc.td]: We can do " compress MOVEBE_rr to BSWAP" in another PR.

File: llvm/lib/Target/X86/X86InstrRAOINT.td (hunk 15 above)
Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrRAOINT.td:34]: BinOpMR_M has no pattern.

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrRAOINT.td]: It's not an issue. You can add a `list<dag> p = []` for it.

File: llvm/lib/Target/X86/X86InstrMisc.td [hunk 17]
```diff
@@ -229,7 +229,7 @@ def PUSHA16  : I<0x60, RawFrm, (outs), (ins), "pushaw", []>,
                OpSize16, Requires<[Not64BitMode]>;
 }
 
-let Constraints = "$src = $dst", SchedRW = [WriteBSWAP32] in {
+let Constraints = "$src = $dst", SchedRW = [WriteBSWAP32], Predicates = [NoNDD_Or_NoMOVBE] in {
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrMisc.td:232]: Why? `Predicates = [NoNDD]` should be sufficient.

Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrMisc.td:232]: Never mind. I found the problem. Add test in llvm/test/CodeGen/X86/bswap.ll

**Review by KanRobert — APPROVED** [review]
LGTM

**Pull request merged**